Add automatic keep-alive for IKEA Trådfri

tbarnekov opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Due to the issues with the UDP stream timing out when there's a NAT device (like Docker in bridge mode) between HA and the IKEA gateway, I've implemented a keep-alive function that queries the gateway every 60 seconds. This will let the NAT device know that the connection is still in use, thus preventing it from discarding it. When the NAT device discards the connection due to inactivity, HA will no longer receive state updates from the gateway. 

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->
The keep-alive functionality basically makes sure a NAT device's connection tracking is updated every minute by sending a request to the gateway and receiving the response. It works similarly to the work-arounds described in #14386 but doesn't require user configuration.

- This PR fixes or closes issue: fixes #14386
- This PR is related to issue: #
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

tbarnekov committed:
Added automatic keep-alive for IKEA Trådfri

MartinHjelmare commented on homeassistant/components/tradfri/__init__.py:
```diff
@@ -147,10 +154,21 @@ async def on_hass_stop(event):
+            await api(gateway.get_gateway_info())
+        except RequestError:
+            _LOGGER.error("Keep-alive failed")
+            await factory.shutdown()
```
I'm not sure we should shutdown the factory. We can't recover from that unless we create a new factory.

The factory will reset the protocol upon error before raising it to us, so I think we can just log the error and keep going.

tbarnekov replied:
Cool, thanks!
I think you're right, and I've pushed the change to the code.

tbarnekov committed:
Don't shut down factory on keep-alive error

MartinHjelmare commented:
Please test this PR for a day or two with the latest change before we merge.

How long did you test the PR before opening it? Are all problems resolved?

tbarnekov commented:
The code was written today, so it's still pretty fresh. I've been monitoring the traffic with Wireshark, and it looks sane so far. I've just updated my HA install with the new code, and will let it soak for a few days.

Is there anything in particular I should be looking out for? Besides error messages in the log of course.

MartinHjelmare commented:
Nothing in particular. Just that the problem is gone as expected and that no other new problems are seen.

hfurubotten commented:
I usually need to reload the IKEA integration once a day here because of these disconnects, so I have also added this to my HA. Will report back any findings I come over with this included in my system. 

One initial thing I found, that might not be that big of a deal, but at least it is mentioned, is that during shutdown the keep-alive might try to run a requests and fail since the factory is shut down just before and log an error.
```
2020-10-14 00:14:29 ERROR (MainThread) [homeassistant.components.tradfri] Keep-alive failed
```
Have restarted a few times since I added this and only seen it once, so it only seem to trigger on unlucky timing.

tbarnekov commented:
> 
> 
> I usually need to reload the IKEA integration once a day here because of these disconnects, so I have also added this to my HA. Will report back any findings I come over with this included in my system.
> 
> One initial thing I found, that might not be that big of a deal, but at least it is mentioned, is that during shutdown the keep-alive might try to run a requests and fail since the factory is shut down just before and log an error.
> 
> ```
> 2020-10-14 00:14:29 ERROR (MainThread) [homeassistant.components.tradfri] Keep-alive failed
> ```
> 
> Have restarted a few times since I added this and only seen it once, so it only seem to trigger on unlucky timing.

Thank you for your feedback.

I agree that there seems to be a possible race-condition during shutdown. Even though it's not "dangerous", I think it should be fixed. I'll see if I can get a fix ready later today.

I slightly skeptical that the keep-alive will fix the disconnects you're having. NAT devices will typically remove the connection after 180 seconds in inactivity, so if you're only seeing disconnects once or twice per day, you either have a lot of activity between HA and the gateway, or the problems are caused by something else.

Anyway, I'm looking forward to an update once you've had it running for a little longer.

balloob commented on homeassistant/components/tradfri/__init__.py:
```diff
@@ -147,10 +154,20 @@ async def on_hass_stop(event):
         )
 
+    async def async_keep_alive(now):
+        try:
```
To avoid the keep-alive after shutdown, add this check:

```suggestion
        if hass.is_stopping:
            return

        try:
```

balloob reviewed (approved):
Looks good! Can be merged once testing shows it to be useful.

tbarnekov committed:
Avoid keep-alive after shutdown

Co-authored-by: Paulus Schoutsen <paulus@home-assistant.io>

hfurubotten commented:
> I slightly skeptical that the keep-alive will fix the disconnects you're having. NAT devices will typically remove the connection after 180 seconds in inactivity, so if you're only seeing disconnects once or twice per day, you either have a lot of activity between HA and the gateway, or the problems are caused by something else.

Your skepticism is completely understandable, have been skeptical many times myself about the issues I have with Tradfri. Haven't completely ruled out that it might be other things interfering with the connection either, but with a lot of testing and different kinds of "keep them talking" solutions (with varying degree of success) leads me to somewhere a UDP timeout occurs and they can't talk to each other anymore. 

I also have the Flux integration active on the IKEA lights, so I think this is the reason I only need to reload it about once a day (sometimes more tho), but haven't tested turning Flux off for quite a while. There is always a light on during the day so that should generate traffic towards the Gateway about every 82 seconds with new warmth settings, with the configuration I have. But.. all IKEA lights are off during the nights, so it has become a morning ritual to reload Trådfri now. 😄

The good news tho is that I didn't need to reload this morning. Has happened before that somehow it magically just still work one day to be disappointed the next, so will still keep an eye on it, but looks promising. There is also no log messages about failed requests towards the Gateway either.

tbarnekov commented:
@hfurubotten Wow, that sounds really promising so far. I’ll keep my fingers crossed that it’ll keep working. I’m looking forward to another update from you tomorrow:)

I only have the Fyrtur blinds, so I only get traffic from the gateway when they move up in the morning and down in the evening. So the integration basically stops working 180 seconds after a reload. I haven’t had a single issue with missing updates so far, and no messages in the log either. 

A big thank you to @balloob who provided a simple fix for the race-condition. It was a lot simpler than the fix I was working on myself 👍 I’ve never coded in Python before, and I only started using HA last week, so there’s a steep learning curve. I appreciate all the help I can get :)

hfurubotten commented:
Busy day today, so didn't have time to check everything for a status update until way out in the evening/night. This morning everything still worked as expected and no reload was necessary, so I could just enjoy all the morning automations without having to kickstart some of them, so really good news! 😀 Has worked like a charm throughout the day too. The log is also clean, no failed requests reported.

Still going strong on your end as well, @tbarnekov ?

tbarnekov commented:
Thank you for the update. Nothing to report on my end. It’s been running like clockwork without needing a reload and no messages in the log.

I think we can merge it for the next release if @MartinHjelmare agrees?

MartinHjelmare commented:
Great!
